### DISCUSSION TITLE
Avoid turbo circular dependency error while using `npm run dev` in non-monorepo (polyrepo) setup

### DISCUSSION THREAD

**Discussion by gnowland** [vercel/turborepo#8200] [Show and tell]
This might be an anti-pattern, but I wanted to add Turbo to my NextJS application in a non-monorepo setup and have it run from `npm run dev` so I don't have to retrain the other contributors (and our CI).

At first Turbo was complaining that I'd created a recursive loop:
```sh
   × root task //#dev (turbo dev) looks like it invokes turbo and might cause a loop
```

To avoid this I added a conditional to my `package.json` `dev` script to check if Turbo was executing it (by checking for the existince of the `TURBO_HASH` environment variable) to run `next dev` if it was, and moved the `turbo dev` call to a `dev:turbo` script since it appears `turbo dev` is parsing the `dev` script text for the words `turbo dev` to throw the above error.

### directory structure
```txt
./
 | src/[...]
 | package.json 
 | turbo.json
```

### `package.json`:
```diff
{
   ...
   "scripts": {
-     "dev": "next dev",
+     "dev": "if test \"$TURBO_HASH\"; then next dev; else npm run dev:turbo; fi",
+     "dev:turbo": "turbo dev",
      ...
   }
   ...
}
```

Now running `npm run dev` runs turbo! 🥳

This pattern can be applied to `build`, `start`, `lint`, etc. 

Hope this helps someone else!

**Comment by luwes** [comment]
I'm facing a similar issue when using the root package as a published NPM package so my NPM workspaces has a workspace pointing to the root `["."]`. It seems valid in NPM workspaces so it would be great if Turbo would be smart enough to not re-run the NPM script that triggered the whole build, or have some exclude scripts from this workspace option.

For now I removed the NPM scripts from the root package and documented to use the `turbo` CLI directly.
https://github.com/muxinc/player.style/pull/7

**Comment by gnowland** [comment]
I also just realized turbo engine, turborepo, and turbopack are for different implementations and I should probably just use turbopack via `"dev": "next dev --turbo"` instead of `turbo dev`. The "turbo-" naming scheme threw me.
